go/mcap/indexed_message_iterator.go

```diff
@@ -254,60 +281,99 @@ func readRecord(r io.Reader) (OpCode, []byte, error) {
 				}
 				if opcode != OpMetadata {
-					return nil, nil, nil, fmt.Errorf("expected metadata record, found %v", data)
+					return nil, fmt.Errorf("expected metadata record, found %v", data)
 				}
 				metadata, err := ParseMetadata(data)
 				if err != nil {
-					return nil, nil, nil, fmt.Errorf("failed to parse metadata record: %w", err)
+					return nil, fmt.Errorf("failed to parse metadata record: %w", err)
 				}
 				err = it.metadataCallback(metadata)
 				if err != nil {
-					return nil, nil, nil, fmt.Errorf("metadata callback failed: %w", err)
+					return nil, fmt.Errorf("metadata callback failed: %w", err)
 				}
 			}
 		}
 	}
 
-	for it.indexHeap.Len() > 0 {
-		ri, err := it.indexHeap.HeapPop()
+	for it.indexHeap.len() > 0 {
+		ri, err := it.indexHeap.Pop()
 		if err != nil {
-			return nil, nil, nil, err
+			return nil, err
 		}
-		if ri.messageIndexEntry == nil {
+		if ri.ChunkSlotIndex == -1 {
 			err := it.loadChunk(ri.chunkIndex)
 			if err != nil {
-				return nil, nil, nil, err
+				return nil, err
 			}
 			continue
 		}
-		chunkOffset := ri.messageIndexEntry.Offset
-		length := binary.LittleEndian.Uint64(ri.buf[chunkOffset+1:])
-		messageData := ri.buf[chunkOffset+1+8 : chunkOffset+1+8+length]
-		message, err := ParseMessage(messageData)
-		if err != nil {
-			return nil, nil, nil, err
+		chunkOffset := ri.MessageOffsetInChunk
+		decompressedChunk := it.decompressedChunks[ri.ChunkSlotIndex]
+		length := binary.LittleEndian.Uint64(decompressedChunk.buf[chunkOffset+1:])
+		messageData := decompressedChunk.buf[chunkOffset+1+8 : chunkOffset+1+8+length]
+		if cap(buf) < len(messageData) {
+			buf = append([]byte{}, messageData...)
+		} else {
+			buf = buf[:len(messageData)]
+			copy(buf, messageData)
 		}
-		channel := it.channels[message.ChannelID]
-		schema := it.schemas[channel.SchemaID]
-		return schema, channel, message, nil
+		if err := msg.PopulateFrom(buf); err != nil {
+			return nil, err
+		}
+		it.decompressedChunks[ri.ChunkSlotIndex].unreadMessages--
+		return buf, nil
+	}
+	return buf, io.EOF
+}
+
+func (it *indexedMessageIterator) Next(buf []byte) (*Schema, *Channel, *Message, error) {
+	msg := &Message{}
+	_, err := it.ReadNextInto(buf, msg)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	schema, channel, err := it.SchemaAndChannelForID(msg.ChannelID)
+	if err != nil {
+		return nil, nil, nil, err
 	}
-	return nil, nil, nil, io.EOF
+	return schema, channel, msg, nil
+}
+
+func (it *indexedMessageIterator) SchemaAndChannelForID(channelID uint16) (*Schema, *Channel, error) {
+	channel, ok := it.channels[channelID]
+	if !ok {
+		return nil, nil, fmt.Errorf("no channel for id %d", channelID)
+	}
+	if channel.SchemaID == 0 {
+		return nil, channel, nil
+	}
+	schema, ok := it.schemas[channel.SchemaID]
+	if !ok {
+		return nil, nil, fmt.Errorf("no associated schema with id %d for channel %d", channel.SchemaID, channelID)
+	}
+	return schema, channel, nil
+}
```

Comment:
isn't this just pushing the same lookups we were previously doing, onto the user? i.e to actually use this code they need to call this on each iteration, thus obviating any benefits of removing it from the signature?